File: frontend/src/types/portfolio.ts
export interface Position {
  symbol: string;
  name: string;
  qty: number;
  side: "short" | "long";
  avgPrice: number;
  currentPrice: number;
  pnl: number;
  pnlPct: number;
  strategy: string;
};

export type StrategyState = "running" | "paused" | "stopped";

export interface Strategy {
  id: string;
  name: string;
  state: StrategyState;
  currentWeight: number;
  targetWeight: number;
  positionsCount: number;
  lastRun: string;
};

export interface SectorAllocationItem {
  sector: string;
  percent: number;
  value: number;
};

export interface EquityPoint {
  date: string;
  value: number;
};

export interface DrawdownPoint {
  date: string;
  value: number;
};

export type OrderStatus = "filled" | "partial" | "cancelled" | "pending"; 
export interface Order {
  id: number;
  time: string;
  type: "BUY" | "SELL";
  symbol: string;
  qty: number;
  status: OrderStatus;
  strategy: string;
};

export type AlertLevel = "warning" | "error";

export interface AlertItem {
  id: number;
  time: string;
  level: AlertLevel;
  message: string;
  strategy: string;
};
export interface BotRun {
  id: number;
  time: string;
  status: "success" | "failed";
  duration: string;
  trades: number;
};

export type Env = "paper" | "live";
export type Range = "1W" | "1M" | "3M" | "1Y" | "ALL";

export type PortfolioSummaryResponse = {
  env: Env;
  as_of: string;
  mode: { environment: Env; kill_switch: boolean };
  status: {
    broker: { state: "up" | "down"; latency_ms?: number };
    worker: { state: "running" | "stopped" | "unknown"; last_heartbeat_at?: string };
    data: { state: "ok" | "lagging" | "down"; lag_seconds?: number };
  };
  account: {
    equity: number;
    cash: number;
    buying_power: number;
    today_pnl: { value: number; pct: number };
    open_positions: { count: number; long: number; short: number };
  };
  exposure: {
    net_pct: number;
    gross_pct: number;
    cash_pct: number;
    top5_concentration_pct: number;
  };
};

export type PortfolioPerformanceResponse = {
  env: Env;
  range: Range;
  benchmark?: string;
  as_of: string;
  equity_curve: { t: string; equity: number }[];
  benchmark_curve?: { t: string; price: number }[];
};

export type PortfolioDrawdownResponse = {
  env: Env;
  range: Range;
  as_of: string;
  drawdown_curve: { t: string; drawdown_pct: number }[];
  summary: { current_drawdown_pct: number; max_drawdown_pct: number };
};

export type PortfolioKpiResponse = {
  env: Env;
  range: Range;
  as_of: string;
  kpi: {
    period_return_pct: number;
    cagr_pct: number;
    volatility_pct: number;
    sharpe: number;
    max_drawdown_pct: number;
    win_rate_pct: number;
  };
  notes?: { risk_free_rate_pct?: number; method?: string };
};

export type PortfolioPositionsResponse = {
  env: Env;
  as_of: string;
  items: Array<{
    symbol: string;
    qty: number;
    side: "long" | "short";
    avg_entry_price: number;
    current_price: number;
    market_value: number;
    unrealized_pnl: { value: number; pct: number };
    strategy?: { user_strategy_id: string; name: string };
  }>;
};

export type PortfolioAllocationResponse = {
  env: Env;
  as_of: string;
  by_sector: Array<{ sector: string; value: number; pct: number; pnl_value?: number }>;
  by_strategy: Array<{ user_strategy_id: string; name: string; value: number; pct: number; pnl_value?: number }>;
  exposure: {
    net_pct: number;
    gross_pct: number;
    cash_pct: number;
    top5_concentration_pct: number;
  };
};

export type PortfolioActivityResponse = {
  env: Env;
  items: Array<{
    type: "order" | "trade" | "alert" | "bot_run";
    id: string;
    t: string;
    data: any;
  }>;
  next_cursor?: string;
};

export type PortfolioAttributionResponse = {
  env: Env;
  range?: Range;
  by: "strategy" | "sector";
  as_of: string;
  items: Array<{
    key: string;
    label: string;
    exposure_pct: number;
    unrealized_pnl_value: number;
    period_contribution_pct?: number;
  }>;
};

export type PortfolioRebalanceRequest = {
  mode: "dry_run" | "execute";
  target_source: "combined" | "strategy";
  strategy_ids?: string[];
  target_weights?: Record<string, number>;
  target_cash_pct?: number;
  overrides?: { cash_buffer?: number };
};

export type PortfolioRebalanceResponse = {
  env: Env;
  mode: "dry_run" | "execute";
  rebalance_id: string;
  status: "preview" | "submitted";
  orders: Array<{
    symbol: string;
    side: "buy" | "sell";
    qty: number;
    notional: number;
    estimated_price: number;
  }>;
};

export type UserStrategyListItem = {
  user_strategy_id: string;
  name: string;
  public_strategy_id?: string;
  state: "running" | "paused" | "stopped";
  positions_count: number;
  today_pnl?: { value: number; pct: number };
  last_run_at?: string;
  params?: Record<string, any>;
  risk_limits?: Record<string, any>;
};

export type UserStrategiesResponse = {
  env: Env;
  items: UserStrategyListItem[];
};

export type UserStrategyDetailResponse = {
  env: Env;
  user_strategy_id: string;
  name: string;
  state: "running" | "paused" | "stopped";
  public_strategy: {
    public_strategy_id: string;
    name: string;
    one_liner?: string;
    param_schema?: Record<
      string,
      { type: "int" | "float" | "string" | "bool"; default?: any; min?: number; max?: number }
    >;
  };
  params: Record<string, any>;
  risk_limits: Record<string, any>;
};

File: backend/app/schemas/portfolio.py
from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


EnvLiteral = Literal["paper", "live"]
RangeLiteral = Literal["1W", "1M", "3M", "1Y", "ALL"]


class PnlBlock(BaseModel):
    model_config = ConfigDict(extra="forbid")

    value: float
    pct: float


class ModeBlock(BaseModel):
    model_config = ConfigDict(extra="forbid")

    environment: EnvLiteral
    kill_switch: bool


class BrokerStatus(BaseModel):
    model_config = ConfigDict(extra="forbid")

    state: Literal["up", "down", "degraded"]
    latency_ms: int


class WorkerStatus(BaseModel):
    model_config = ConfigDict(extra="forbid")

    state: Literal["running", "stopped", "error", "queued"]
    last_heartbeat_at: str


class StatusBlock(BaseModel):
    model_config = ConfigDict(extra="forbid")

    broker: BrokerStatus
    worker: WorkerStatus


class OpenPositionsBlock(BaseModel):
    model_config = ConfigDict(extra="forbid")

    count: int
    long: int
    short: int


class AccountBlock(BaseModel):
    model_config = ConfigDict(extra="forbid")

    equity: float
    cash: float
    buying_power: float
    today_pnl: PnlBlock
    open_positions: OpenPositionsBlock


class ExposureBlock(BaseModel):
    model_config = ConfigDict(extra="forbid")

    net_pct: float
    gross_pct: float
    cash_pct: float
    top5_concentration_pct: float


class PortfolioSummaryResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    as_of: str
    mode: ModeBlock
    status: StatusBlock
    account: AccountBlock
    exposure: ExposureBlock


class PortfolioOverviewResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    as_of: str
    summary: PortfolioSummaryResponse
    allocation: "PortfolioAllocationResponse"


class EquityPoint(BaseModel):
    model_config = ConfigDict(extra="forbid")

    t: str
    equity: float


class BenchmarkPoint(BaseModel):
    model_config = ConfigDict(extra="forbid")

    t: str
    price: float


class PortfolioPerformanceResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    range: RangeLiteral
    benchmark: str
    as_of: str
    equity_curve: List[EquityPoint]
    benchmark_curve: List[BenchmarkPoint]


class DrawdownPoint(BaseModel):
    model_config = ConfigDict(extra="forbid")

    t: str
    drawdown_pct: float


class DrawdownSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    current_drawdown_pct: float
    max_drawdown_pct: float


class PortfolioDrawdownResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    range: RangeLiteral
    drawdown_curve: List[DrawdownPoint]
    summary: DrawdownSummary


class PortfolioKpi(BaseModel):
    model_config = ConfigDict(extra="forbid")

    period_return_pct: float
    cagr_pct: float
    volatility_pct: float
    sharpe: float
    max_drawdown_pct: float
    win_rate_pct: float


class PortfolioKpiResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    range: RangeLiteral
    kpi: PortfolioKpi


class PositionStrategyRef(BaseModel):
    model_config = ConfigDict(extra="forbid")

    user_strategy_id: str
    name: str


class PositionItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    symbol: str
    qty: float
    side: Literal["long", "short"]
    avg_entry_price: float
    current_price: float
    market_value: float
    unrealized_pnl: PnlBlock
    strategy: PositionStrategyRef


class PortfolioPositionsResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    items: List[PositionItem]


class SectorAllocation(BaseModel):
    model_config = ConfigDict(extra="forbid")

    sector: str
    value: float
    pct: float
    pnl_value: float


class StrategyAllocation(BaseModel):
    model_config = ConfigDict(extra="forbid")

    user_strategy_id: str
    name: str
    value: float
    pct: float
    pnl_value: float


class PortfolioAllocationResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    by_sector: List[SectorAllocation]
    by_strategy: List[StrategyAllocation]
    exposure: ExposureBlock


class AttributionItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    key: str
    label: str
    exposure_pct: float
    unrealized_pnl_value: float
    period_contribution_pct: float


class PortfolioAttributionResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    range: Optional[RangeLiteral] = None
    by: Literal["strategy", "sector"]
    items: List[AttributionItem]


class RiskLimits(BaseModel):
    model_config = ConfigDict(extra="forbid")

    max_weight_per_asset: float
    cash_buffer: float
    max_turnover_pct: Optional[float] = None


class UserStrategySummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    user_strategy_id: str
    name: str
    public_strategy_id: str
    state: Literal["running", "paused", "stopped", "idle", "error"]
    positions_count: int
    today_pnl: PnlBlock
    last_run_at: str
    params: Dict[str, Any]
    risk_limits: RiskLimits


class UserStrategiesResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    items: List[UserStrategySummary]


class PublicStrategyBrief(BaseModel):
    model_config = ConfigDict(extra="forbid")

    public_strategy_id: str
    name: str
    one_liner: str
    param_schema: Dict[str, Any]


class StrategyRunSummary(BaseModel):
    model_config = ConfigDict(extra="forbid")

    run_id: str
    started_at: str
    status: Literal["success", "failed", "running", "queued"]
    orders_created: int


class UserStrategyDetailResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    user_strategy_id: str
    name: str
    state: Literal["running", "paused", "stopped", "idle", "error"]
    public_strategy: PublicStrategyBrief
    params: Dict[str, Any]
    risk_limits: RiskLimits
    recent_runs: List[StrategyRunSummary]


class UpdateUserStrategyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: Optional[str] = None
    params: Optional[Dict[str, Any]] = None
    risk_limits: Optional[RiskLimits] = None


class UpdateUserStrategyResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    ok: bool
    user_strategy_id: str
    updated_at: str


class StrategyStateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    action: Literal["start", "pause", "stop"]


class StrategyStateResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    ok: bool
    user_strategy_id: str
    state: Literal["running", "paused", "stopped"]


class RebalanceOverrides(BaseModel):
    model_config = ConfigDict(extra="forbid")

    cash_buffer: Optional[float] = None


class PortfolioRebalanceRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    mode: Literal["dry_run", "execute"]
    target_source: Literal["combined", "strategy"]
    strategy_ids: Optional[List[str]] = None
    target_weights: Optional[Dict[str, float]] = None
    target_cash_pct: Optional[float] = Field(default=None, ge=0, le=100)
    allow_new_positions: Optional[bool] = False
    overrides: Optional[RebalanceOverrides] = None


class RebalanceOrderPreview(BaseModel):
    model_config = ConfigDict(extra="forbid")

    symbol: str
    side: Literal["buy", "sell"]
    qty: float
    notional: float
    estimated_price: float


class RebalanceSubmitResult(BaseModel):
    model_config = ConfigDict(extra="forbid")

    symbol: str
    side: Literal["buy", "sell"]
    qty: float
    notional: float
    estimated_price: float
    order_id: Optional[str] = None
    status: Literal["submitted", "failed"]
    error: Optional[str] = None


class PortfolioRebalanceResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    mode: Literal["dry_run", "execute"]
    rebalance_id: str
    status: Literal["preview", "submitted"]
    orders: List[RebalanceOrderPreview]
    submitted: Optional[List[RebalanceSubmitResult]] = None
    alpaca_positions: Optional[List[PositionItem]] = None


class RebalanceTargetItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    strategy_id: str
    target_weight_pct: float
    target_cash_pct: float
    updated_at: str


class PortfolioRebalanceTargetsResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    items: List[RebalanceTargetItem]


class ActivityItem(BaseModel):
    model_config = ConfigDict(extra="forbid")

    type: Literal["order", "trade", "alert", "bot_run"]
    id: str
    t: str
    data: Dict[str, Any]


class PortfolioActivityResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    env: EnvLiteral
    items: List[ActivityItem]
    next_cursor: Optional[str] = None


PortfolioOverviewResponse.model_rebuild()
